lib/Differentiator/ReverseModeVisitor.cpp

```diff
@@ -584,68 +584,144 @@ namespace clad {
   }
 
   void ReverseModeVisitor::DifferentiateWithEnzyme() {
-    // FIXME: Generalize this function to differentiate other kinds
-    // of function prototypes
     unsigned numParams = m_Function->getNumParams();
     auto origParams = m_Function->parameters();
     llvm::ArrayRef<ParmVarDecl*> paramsRef = m_Derivative->parameters();
     auto originalFnType = dyn_cast<FunctionProtoType>(m_Function->getType());
 
-    // Case 1: The function to be differentiated is of type double
-    // func(double* arr){...};
-    //        or double func(double arr[n]){...};
-    if (numParams == 1) {
-      QualType origTy = origParams[0]->getOriginalType();
-      if (origTy->isConstantArrayType() || origTy->isPointerType()) {
-        // Extract Pointer from Clad Array Ref
-        auto arrayRefNameExpr = BuildDeclRef(paramsRef[1]);
-        auto getPointerExpr = BuildCallExprToMemFn(arrayRefNameExpr, "ptr", {});
-        auto arrayRefToArrayStmt = BuildVarDecl(
-            origTy, "d_" + paramsRef[0]->getNameAsString(), getPointerExpr);
-        addToCurrentBlock(BuildDeclStmt(arrayRefToArrayStmt),
-                          direction::forward);
+    // Extract Pointer from Clad Array Ref
+    llvm::SmallVector<VarDecl*, 8> cladRefParams;
+    for (int i = 0; i < numParams; i++) {
+      QualType paramType = origParams[i]->getOriginalType();
+      QualType q2(paramType->getAs<PointerType>(), 0);
+      if (!(paramType->isArrayType() || paramType->isPointerType())) {
```

Comment:
```suggestion
      if (!isArrayOrPointerType(origParams[i]->getOriginalType())) {
```
We might want to leverage `isDifferentiableType`?

Reply:
No, `isDifferentiableType()` will return true if the QualType is real or pointer/array types. Here I explicitly want to check if it is realType.  Maybe I should rephrase this to `isRealType(origParams[i]->getOriginalType())`

Reply:
also I cant get rid of the paramType variable, because I am using in many more places from here, but I have removed q2

Reply:
I think using `isArrayOrPointerType` is fine and keeping `paramType` is also fine.